packages/moonstone/VirtualList/VirtualListBase.js

```diff
@@ -420,58 +420,38 @@ const VirtualListBaseFactory = (type) => {
 			const {cbScrollTo, spacing, wrap} = this.props;
-			const {dimensionToExtent, primary: {clientSize, gridSize}, scrollPosition} = this.uiRefCurrent;
+			const {dimensionToExtent, primary: {clientSize, gridSize}, scrollPositionTarget} = this.uiRefCurrent;
 			const index = getNumberValue(target.dataset.index);
 
 			this.isScrolledBy5way = false;
 			this.isScrolledByJump = false;
 
 			if (nextIndex >= 0) {
-				const {dataSize} = this.props;
-				const column = index % dimensionToExtent;
-				const row = (index - column) % dataSize / dimensionToExtent;
-				const nextColumn = nextIndex % dimensionToExtent;
-				const nextRow = (nextIndex - nextColumn) % dataSize / dimensionToExtent;
-				let
-					isNextItemInView = false,
-					isCurrentItemInView = false,
-					numOfItemsInPage;
+				const
+					row = Math.floor(index / dimensionToExtent),
+					nextRow = Math.floor(nextIndex / dimensionToExtent);
+				let isNextItemInView = false;
 
 				if (this.props.itemSizes) {
-					const container = this.uiRefCurrent.containerRef.current;
-					const inView = [index, nextIndex].map((i) => {
-						const node = this.uiRefCurrent.containerRef.current.querySelector(`[data-index='${i}']`);
-
-						if (container && node) {
-							const containerRects = container.getBoundingClientRect();
-							const nodeRects = node.getBoundingClientRect();
-
-							if (
-								nodeRects.x >= containerRects.x &&
-								nodeRects.x + nodeRects.width <= containerRects.x + containerRects.width &&
-								nodeRects.y >= containerRects.y &&
-								nodeRects.y + nodeRects.height <= containerRects.y + containerRects.height
-							) {
-								return true;
-							}
-						}
-
-						return false;
-					});
-					numOfItemsInPage = dimensionToExtent;
-					isCurrentItemInView = inView[0];
-					isNextItemInView = inView[1];
+					isNextItemInView = this.uiRefCurrent.itemPositions[nextIndex].position >= scrollPositionTarget &&
+						this.uiRefCurrent.getItemBottomPosition(nextIndex) <= scrollPositionTarget + clientSize;
 				} else {
-					const firstFullyVisibleIndex = Math.ceil(scrollPosition / gridSize) * dimensionToExtent;
-					numOfItemsInPage = Math.floor((clientSize + spacing) / gridSize) * dimensionToExtent;
-					isCurrentItemInView = index >= firstFullyVisibleIndex && index < firstFullyVisibleIndex + numOfItemsInPage;
+					const
+						firstFullyVisibleIndex = Math.ceil(scrollPositionTarget / gridSize) * dimensionToExtent,
+						numOfItemsInPage = Math.floor((clientSize + spacing) / gridSize) * dimensionToExtent;
 					isNextItemInView = nextIndex >= firstFullyVisibleIndex && nextIndex < firstFullyVisibleIndex + numOfItemsInPage;
 				}
 
 				this.lastFocusedIndex = nextIndex;
 
-				if (isNextItemInView && (isCurrentItemInView || numOfItemsInPage !== dimensionToExtent) || row === nextRow) {
+				if (
+					isNextItemInView ||
+					row === nextRow // Needed when the number of the visible row is 1 and the number of the fully visible row is 0.
```

Comment:
```
Since row === nextRow, then focus will change to the adjacent item (which is partially off-screen), but it will not scroll into view.
```
The case when `row === nextRow` was the one that the number of the fully visible row was 0. But it was not a real case because VirtualGridList should have enough height to show an item. So I removed this condition. Even though I removed this condition, I think that it makes sense that Vertical VirtualGridList scrolls up and down via 5-way key Left and Right to move to the next item horizontally to stick to the top and to the bottom. There is no exact requirement for this case now. So in my opinion, we could go ahead.

```
I see that the value of isNextItemInView can be incorrect if you have a VirtualGridList that contains 2 rows, where the top of row 1 is partially out of view and also the bottom of row 2 is partially out of view.
```
I think we calculated the last fully visible index incorrectly by adding `firstFullyVisibleIndex` and `numOfItemsInPage`. After calculating `lastFullyVisibleIndex` exactly, it works as we expected. 

Now I seems to me that everything works correctly. If starting to use `the currently-focused item` in our condition, then the condition gets complex. So I tried to avoid using it. Could you check anything else that I missed?